File: leetcode/129-sum-root-to-leaf-numbers.py
# https://leetcode.com/problems/sum-root-to-leaf-numbers/

from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def sumNumbers(self, root: Optional[TreeNode], numbers = 0) -> int:
        if not root: return 0
        if not root.left and not root.right:
            return (numbers * 10) + root.val

        numbers = (numbers * 10) + root.val

        return self.sumNumbers(root.left, numbers) + self.sumNumbers(root.right, numbers)

File: leetcode/129-sum-root-to-leaf-numbers.js
// https://leetcode.com/problems/sum-root-to-leaf-numbers/

// Definition for a binary tree node.
function TreeNode(val, left, right) {
  this.val = (val===undefined ? 0 : val)
  this.left = (left===undefined ? null : left)
  this.right = (right===undefined ? null : right)
}

/**
 * @param {TreeNode} root
 * @return {number}
 */
var sumNumbers = function(root, numbers = 0) {
  if (!root) return 0;
  if (!root.left && !root.right) {
    return (numbers * 10) + root.val;
  }

  numbers = (numbers * 10) + root.val;

  return sumNumbers(root.left, numbers) + sumNumbers(root.right, numbers);
};
